sycl/test/check_device_code/matrix/matrix-nvptx-tf32-test.cpp

```diff
@@ -0,0 +1,141 @@
+// validated before supporting any sm versions beyond sm_86. The reason that we
+// choose to use the m16n16k16 instruction is that it allows the significant
+// advantage of being able to use a portable interface across Intel and Nvidia
+// backends.
+
+#include <CL/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+// M, N, K define the sizes of dimensions of the three matrix types (a, b,
+// accumulator) used per subgroup operation.
+constexpr int M = 16; // number of rows of accumulator,
+                      // number of cols of b.
+constexpr int N = 16; // number of cols of accumulator,
+                      // number of rows of a.
+constexpr int K = 8;  // number of cols of a/number of rows of b.
+
+// float is used in this test as the storage type for tf32
+float A[M * K];
+float B[K * N];
+float C[M * N];
+float D[M * N];
+
+int main() {
+
+  buffer<float, 1> bufA(A, range<1>(M * K)); // will be used as tf32
+  buffer<float, 1> bufB(B, range<1>(K * N)); // will be used as tf32
+  buffer<float, 1> bufC(C, range<1>(M * N));
+  buffer<float, 1> bufD(D, range<1>(M * N));
+
+  queue q;
+
+  q.submit([&](handler &cgh) {
+    auto accA = bufA.get_access<access::mode::read_write>(cgh);
+    auto accB = bufB.get_access<access::mode::read_write>(cgh);
+    auto accC = bufC.get_access<access::mode::read_write>(cgh);
+    auto accD = bufD.get_access<access::mode::read_write>(cgh);
+
+    cgh.parallel_for<class row_row>(
+        nd_range<2>({1, 32}, {1, 32}),
+        [=](nd_item<2> item) [[sycl::reqd_work_group_size(1, 1, 32)]] {
+          sycl::sub_group sg = item.get_sub_group();
+
+          joint_matrix<precision::tf32, matrix_use::a, M, K,
+                       matrix_layout::row_major>
+              sub_a;
+
+          joint_matrix<precision::tf32, matrix_use::b, K, N,
+                       matrix_layout::row_major>
+              sub_b;
+
+          joint_matrix<float, matrix_use::accumulator, M, N,
+                       matrix_layout::row_major>
+              sub_c;
+
+          //CHECK: tail call { i32, i32, i32, i32 } @llvm.nvvm.wmma.m16n16k8.load.a.row.stride.tf32.p0i32(i32* %call.ascast.i.i{{.*}}.i, i32 8) #{{.*}}
+          joint_matrix_load(sg, sub_a, accA.get_pointer(), K);
+          //CHECK: tail call { i32, i32, i32, i32 } @llvm.nvvm.wmma.m16n16k8.load.b.row.stride.tf32.p0i32(i32* %call.ascast.i.i{{.*}}.i, i32 16) #{{.*}}
+          joint_matrix_load(sg, sub_b, accB.get_pointer(), N);
+          //CHECK: tail call { float, float, float, float, float, float, float, float } @llvm.nvvm.wmma.m16n16k16.load.c.row.stride.f32.p1f32(float addrspace(1)* %_arg_accC, i32 16) #{{.*}}
+          joint_matrix_load(sg, sub_c, accC.get_pointer(), N);
+
+          // CHECK: tail call i32 @llvm.nvvm.f2tf32.rna(float {{.*}}
+          // Round a, b to tf32
+          for (auto i = 0; i < 4; ++i)
+            sub_a.data[i] = float_to_tf32(sub_a.data[i]);
```

Comment:
this should be the expected way to perform the rounding, if users want to, but I am still find exposing ".data" is different from the element wise indexing we are currently doing.
I would recommend moving from this to the current API:
sub_a.get_wi_data()[i] = float_to_tf32(sub_a.get_wi_data()[i]);

Reply:
This is exactly what we will do: I switched the impl [here](https://github.com/intel/llvm/pull/5964) to use `marray` for `data` in `joint_matrix`: Then `get_wi_data()[i]` will call  `get_wi_elem` that will return the ith element of the `marray`. We will loop over `get_wi_data.length()` as you do too.